### PR TITLE
Added Basic Auth support for mitmweb interface

### PR DESCRIPTION
**Pull request by ganguera**
Added Basic Auth support for mitmweb interface as requested in #802. It supports both single user format ("username:password") or .htpasswd file.

### CODE REVIEW

File: libmproxy/web/app.py [hunk 1]
```diff
@@ -240,10 +277,14 @@ def put(self):
...
-    def __init__(self, master, debug):
+    def __init__(self, master, debug, singleuser, htpasswd):
         self.master = master
+        self.additional_args = dict(
+            singleuser=singleuser,
+            htpasswd=htpasswd,
+        )
         handlers = [
-            (r"/", IndexHandler),
+            (r"/", IndexHandler, self.additional_args),
```
Comment (mhils) [inline libmproxy/web/app.py]: So, this looks like it only implements basic auth for the index page, but not the API?

File: libmproxy/web/app.py [hunk 2]
```diff
@@ -25,6 +27,41 @@ def set_default_headers(self):
...
+                        self.set_auth_headers()
+                        raise APIError(401, "Invalid username or password.")
+                elif self.htpasswd:
+                    try:
+                        password_manager = authentication.PassManHtpasswd(self.htpasswd)
+                    except ValueError as v:
+                        raise APIError(500, "%s. Please contact your systems administrator." % v.message)
+                    if not password_manager.test(self.username, self.password):
+                        self.set_auth_headers()
+                        raise APIError(401, "Invalid username or password.")
```
Comment (mhils) [inline libmproxy/web/app.py]: This looks like there's quite a bit of duplication here. It'd be useful if we could just pass a password_manager to whatever authentication thingie we're using. The `password_manager` could be initialized when we parse the command line parameters.

**Comment by mhils** [comment]
Any progress on this?

**Comment by ganguera** [comment]
Hi @mhils, sorry for the delay. Tonight I'll pull request the refactor.

File: libmproxy/web/app.py (hunk 1 above)
Comment (ganguera) [inline libmproxy/web/app.py]: You're right. I need to think about it.

**Comment by mhils** [comment]
This still looks like it only protects the index page?

**Comment by ganguera** [comment]
You're right @mhils , so far this only protects the index page. I think it's easy to enable the basic auth for the rest of the URLs except "/updates", that are being processed with the "tornado.websocket.WebSocketHandler". I need to check whether is possible to enable basic auth using websockets.

**Comment by ganguera** [comment]
Hi @mhils !

As requested, now I'm parsing the specific web options in the web module. I think I also find a way to authorize the usage of websockets by setting / getting a secure cookie.

Looking forward to hearing from you.

**Comment by mhils** [comment]
Thanks, this looks better now. Let me know once we have authentication for the WebSocket channel :)

**Comment by ganguera** [comment]
Sorry guys, I need help to merge this...

**Pull request closed without merging**